### PR TITLE
Pre-scala 2.13: inline Box

### LINKED ISSUE #24764: box

### PR DESCRIPTION
**Pull request by DavidLawes**
## What does this change?
In advance of updating the scala version we want to update as many dependencies as possible to be compatible with both 2.12 and 2.13.

We tried updating frontend scala to 2.13. One of the compilation errors was from com.gu.box. Given [box](https://github.com/guardian/box) is just a very thin wrapper of `AtomicReference`, and only used by a handful of repos, it's been suggested to inline it. 

In this PR we remove Box as a dependency and define it in `common`. Most of the changes relate to updated imports.

One specific error caused me a bit of confusion:
- Inlining Box resulted in a failing test in [DurationMetricTest](https://github.com/guardian/frontend/blob/dlawes/update-guBox/common/test/metrics/DurationMetricTest.scala#L45) --> the test failed because the list length was not expected (ref: https://teamcity.gutools.co.uk/viewLog.html?buildId=752344&buildTypeId=dotcom_master&tab=buildLog&branch_dotcom=dlawes%2Fupdate-guBox&_focus=15032)
- I could see that [FrontendMetrics](https://github.com/guardian/frontend/blob/dlawes/update-guBox/common/app/metrics/FrontendMetrics.scala#L130) was still attempting to add new data points to the head of the list, but the existing data points could not be resolved (`unbound placeholder parameter error`).
- I noticed an equivalent function in the same file: https://github.com/guardian/frontend/blob/dlawes/update-guBox/common/app/metrics/FrontendMetrics.scala#L162. This was not covered by tests; I added tests and saw it was also failing.

^^ to resolve the failing tests we had to update the methods defined in the Box class. There was actually an error in the library itself (around variable names and scopes), which meant that we were only ever altering or modifying the original value for our ref, never the latest. Now that these methods in the local Box class are updated (or removed, if not being used), our tests pass.

## Does this change need to be reproduced in dotcom-rendering ?

- [x] No
- [ ] Yes (please indicate your plans for DCR Implementation)

## Screenshots

N/A

### Does this affect other platforms?

- [ ] AMP 
- [ ] Apps
- [ ] Other (please specify)

### Does this affect GLabs Paid Content Pages? Should it have support for Paid Content?

- [x] No
- [ ] Yes (please give details)

### Does this change break ad-free?

- [x] No
- [ ] It did, but tests caught it and I fixed it
- [ ] It did, but there was no test coverage so I added that then fixed it

### Does this change update the version of CAPI we're using?

- [x] No, all the existing database files are just fine
- [ ] Yes, and I have [re-run all the tests locally and checked in all the updated data/database/xyz files](https://github.com/guardian/frontend/blob/main/docs/03-dev-howtos/15-updating-test-database.md)

### Accessibility test checklist

- [ ] [Tested with screen reader](https://accessibility.gutools.co.uk/testing/web/screen-readers/)
- [ ] [Navigable with keyboard](https://accessibility.gutools.co.uk/testing/web/keyboard-navigation/)
- [ ] [Colour contrast passed](https://accessibility.gutools.co.uk/testing/web/colour-contrast/)

### Tested

- [ ] Locally
- [ ] On CODE (optional)

### CODE REVIEW

File: common/app/common/Box.scala [hunk 1]
```diff
@@ -0,0 +1,46 @@
+package common
```
Comment (DavidLawes) [inline common/app/common/Box.scala:1]: This is a copypasta from https://github.com/guardian/box/blob/master/src/main/scala/com/gu/Box.scala

File: common/test/common/BoxTest.scala [hunk 2]
```diff
@@ -0,0 +1,16 @@
+package common
```
Comment (DavidLawes) [inline common/test/common/BoxTest.scala:1]: This is copypasta from https://github.com/guardian/box/blob/master/src/test/scala/com/gu/BoxSpec.scala

File: common/test/metrics/DurationMetricTest.scala [hunk 3]
```diff
@@ -44,4 +45,44 @@ class DurationMetricTest extends FlatSpec with Matchers {
     dataPoints.length should be(7)
     dataPoints.splitAt(4)._1 should be(allMetrics)
   }
+
+  "SamplerMetric" should "start off empty" in {
```
Comment (DavidLawes) [inline common/test/metrics/DurationMetricTest.scala:49]: The SamplerMetric functionality is almost identical to DurationMetric count, but the recording of the metric/sample is handled by 2 different functions. These tests cover the SamplerMetric recording functionality, they were missing before and helped me try to debug what was going on

File: common/app/metrics/FrontendMetrics.scala [hunk 4]
```diff
@@ -128,7 +127,7 @@ final case class DurationMetric(override val name: String, override val metricUn
   }
 
   // Public for tests.
-  def record(dataPoint: DurationDataPoint): Unit = dataPoints.alter(dataPoint :: _)
+  def record(dataPoint: DurationDataPoint): Unit = dataPoints.alter(dataPoint :: dataPoints.get())
```
Comment (DavidLawes) [inline common/app/metrics/FrontendMetrics.scala:130]: Without this change I was seeing an 'unbound placeholder parameter error' which I think means scala didn't know how to interpret `_`. What I think we want from this function is to append new data points to the list (without this change we just overwrote any existing datapoints, List always had length of 1).

I don't understand why my change introduced this error though. Any ideas?

File: common/app/metrics/FrontendMetrics.scala [hunk 5]
```diff
@@ -160,7 +159,7 @@ final case class SamplerMetric(override val name: String, override val metricUni
   }
 
   def recordSample(sampleValue: Double, sampleTime: DateTime): Future[List[SampledDataPoint]] =
-    dataPoints.alter(SampledDataPoint(sampleValue, sampleTime) :: _)
+    dataPoints.alter(SampledDataPoint(sampleValue, sampleTime) :: dataPoints.get())
```
Comment (DavidLawes) [inline common/app/metrics/FrontendMetrics.scala:162]: I was curious if this function exhibited the same behaviour as `record` above (i.e. `_` results in an 'unbound placeholder parameter error'). After writing tests I confirmed this and believe we now add new sampled data points to the head of the list.

File: common/app/common/Box.scala [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+  def alter(t: T): Future[T]
+  def alter(t: T => T): Future[T]
+}
+
+object Box {
+  def apply[T](initialValue: T): Box[T] = new AtomicRefBox[T](initialValue)
+}
+
+private class AtomicRefBox[T](initialValue: T) extends Box[T] {
```
Comment (DavidLawes) [inline common/app/common/Box.scala:23]: Note the omission of `modify`, `map` and `flatMap` (https://github.com/guardian/box/blob/master/src/main/scala/com/gu/Box.scala#L38). These functions aren't used in `frontend` and the `modify` function has not been implemented correctly in `box`.

File: common/app/common/Box.scala [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+  private val ref: AtomicReference[T] = new AtomicReference[T](initialValue)
```
Comment (DavidLawes) [inline common/app/common/Box.scala:24]: Note the renaming of the initial value. In `box` this is named [t](https://github.com/guardian/box/blob/master/src/main/scala/com/gu/Box.scala#L28) meaning that in the alter and modify methods `t` was always resolving to this initial value.

Comment (jamesgorrie) [reply common/app/common/Box.scala:24]: This comment and [the above one](https://github.com/guardian/frontend/pull/24768/files#r830926884) are good comments, would the be useful to future devs coming to this code?

**Review by jamesgorrie — APPROVED** [review]

File: common/app/common/Box.scala (hunk 7 above)
Comment (DavidLawes) [reply common/app/common/Box.scala:30]: Good shout, will add :)

**Pull request merged**